library/core/resource-loader/src/main/java/org/quiltmc/qsl/resource/loader/impl/ResourceLoaderImpl.java

```diff
@@ -0,0 +1,325 @@
+							+ " already in resource reloader set!"
+			);
+		}
+	}
+
+	/**
+	 * Sorts the given resource reloaders to satisfy dependencies.
+	 *
+	 * @param reloaders the resource reloaders to sort
+	 */
+	private void sort(List<ResourceReloader> reloaders) {
+		reloaders.removeAll(this.addedReloaders);
+
+		// General rules:
+		// - We *do not* touch the ordering of vanilla listeners. Ever.
+		//   While dependency values are provided where possible, we cannot
+		//   trust them 100%. Only code doesn't lie.
+		// - We addReloadListener all custom listeners after vanilla listeners. Same reasons.
+
+		var reloadersToAdd = new ArrayList<>(this.addedReloaders);
+		var resolvedIds = new HashSet<Identifier>();
+
+		for (var reloader : reloaders) {
+			if (reloader instanceof IdentifiableResourceReloader identifiableResourceReloader) {
+				resolvedIds.add(identifiableResourceReloader.getQuiltId());
+			}
+		}
+
+		int lastSize = -1;
+
+		while (reloaders.size() != lastSize) {
+			lastSize = reloaders.size();
+
+			Iterator<IdentifiableResourceReloader> it = reloadersToAdd.iterator();
+
+			while (it.hasNext()) {
+				IdentifiableResourceReloader listener = it.next();
+
+				if (resolvedIds.containsAll(listener.getQuiltDependencies())) {
+					resolvedIds.add(listener.getQuiltId());
+					reloaders.add(listener);
+					it.remove();
+				}
+			}
+		}
+
+		for (var reloader : reloadersToAdd) {
+			LOGGER.warn("Could not resolve dependencies for resource reloader: " + reloader.getQuiltId() + "!");
+		}
+	}
+
+	/* Default resource pack stuff */
+
+	private static Path locateDefaultResourcePack(ResourceType type) {
+		try {
+			// Locate MC jar by finding the URL that contains the assets root.
+			URL assetsRootUrl = DefaultResourcePack.class.getResource("/" + type.getDirectory() + "/.mcassetsroot");
+
+			return Paths.get(assetsRootUrl.toURI()).resolve("../..").toAbsolutePath().normalize();
+		} catch (Exception exception) {
+			throw new RuntimeException("Quilt: Failed to locate Minecraft assets root!", exception);
+		}
+	}
+
+	public static ModNioResourcePack locateAndLoadDefaultResourcePack(ResourceType type) {
+		return new ModNioResourcePack(
+				FabricLoader.getInstance().getModContainer("minecraft").map(ModContainer::getMetadata).orElseThrow(),
+				locateDefaultResourcePack(type),
+				type,
+				() -> {
+				},
+				ResourcePackActivationType.ALWAYS_ENABLED
+		);
+	}
+
+	/* Mod resource pack stuff */
+
+	/**
+	 * Appends mod resource packs to the given list.
+	 *
+	 * @param packs   the resource pack list to append
+	 * @param type    the type of resource
+	 * @param subPath the resource pack sub path directory in mods, may be {@code null}
+	 */
+	public static void appendModResourcePacks(List<ResourcePack> packs, ResourceType type, @Nullable String subPath) {
+		for (var container : FabricLoader.getInstance().getAllMods()) {
+			if (container.getMetadata().getType().equals("builtin")) {
```

Comment:
Not sure this api is going to live in loader, I'm guessing you need this anyways.